dartagnan/src/main/java/com/dat3m/dartagnan/wmm/axiom/Acyclic.java

```diff
@@ -126,20 +126,37 @@ public BooleanFormula consistent(SolverContext ctx) {
 		BooleanFormulaManager bmgr = fmgr.getBooleanFormulaManager();
         IntegerFormulaManager imgr = fmgr.getIntegerFormulaManager();
 
-        BooleanFormula enc = bmgr.makeTrue();
-        for(Tuple tuple : rel.getEncodeTupleSet()){
-            Event e1 = tuple.getFirst();
-            Event e2 = tuple.getSecond();
-			enc = bmgr.and(enc, bmgr.implication(rel.getSMTVar(tuple, ctx), 
-									imgr.lessThan(
-											Utils.intVar(rel.getName(), e1, ctx), 
-											Utils.intVar(rel.getName(), e2, ctx))));
-        }
+        BooleanFormula enc = bmgr.makeTrue(); 
+        if(negated) {
+        	// We use Boolean variables which guess the edges and nodes constituting the cycle. 
+        	// We ensure that for every event in the cycle, there should be at least one incoming 
+        	// edge and at least one outgoing edge that are also in the cycle.
+            for(Event e : rel.getEncodeTupleSet().stream().map(t -> t.getFirst()).collect(Collectors.toSet())){
+            	BooleanFormula in = bmgr.makeFalse();
+            	for(Tuple pre : rel.getEncodeTupleSet().getBySecond(e)) {
+            		in = bmgr.or(in, rel.getSMTCycleVar(pre, ctx));
+            	}
+            	BooleanFormula out = bmgr.makeFalse();
+            	for(Tuple post : rel.getEncodeTupleSet().getByFirst(e)) {
+            		out = bmgr.or(out, rel.getSMTCycleVar(post, ctx));
+            	}
+            	enc = bmgr.and(enc, bmgr.implication(Utils.cycleVar(rel.getName(), e, ctx), bmgr.and(in , out)));
```

Comment:
The meaning of `rel.getSMTCycleVar` is never encoded (i.e. its derivation rules). 
Shouldn't you just use `rel.getSMTVar` here?

Reply:
The rest of the cycle encoding was in the next page of the porthos paper and I missed it XD.

> Shouldn't you just use rel.getSMTVar here?

No, that one should refer to the cycle edges. The new parts refer to the original relation edge.